Sources/SwiftRefactor/CallToTrailingClosures.swift

```diff
@@ -0,0 +1,131 @@
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftBasicFormat
+import SwiftSyntax
+
+/// Convert a call with inline closures to one that uses trailing closure
+/// syntax. Returns `nil` if there's already trailing closures or there are no
+/// closures within the call. Pass `conversionStart` to specify an index to
+/// start the conversion from, ie. to skip converting closures before
+/// `conversionStart`
+///
+/// ## Before
+/// ```
+/// someCall(closure1: { arg in
+///   return 1
+/// }, closure2: { arg in
+///   return 2
+/// })
+/// ```
+///
+/// ## After
+/// ```
+/// someCall { arg in
+///   return 1
+/// } closure2: { arg in
+///   return 2
+/// }
+/// ```
+public struct CallToTrailingClosures: SyntaxRefactoringProvider {
+  public struct Options {
+    public let startAt: Int
+
+    public init(startAt: Int = 0) {
+      self.startAt = startAt
+    }
+  }
+  public typealias Context = Options
+
+  // TODO: Rather than returning nil, we should consider throwing errors with
+  // appropriate messages instead.
+  public static func refactor(syntax call: FunctionCallExprSyntax, in context: Context = Context()) -> FunctionCallExprSyntax? {
+    return call.convertToTrailingClosures(from: context.startAt)?.formatted().as(FunctionCallExprSyntax.self)
+  }
+}
+
+extension FunctionCallExprSyntax {
+  fileprivate func convertToTrailingClosures(from startAt: Int) -> FunctionCallExprSyntax? {
+    guard trailingClosure == nil, additionalTrailingClosures == nil, leftParen != nil, rightParen != nil else {
+      // Already have trailing closures
+      return nil
+    }
+
+    var closures = [(label: TokenSyntax, closure: ClosureExprSyntax)]()
+    for arg in argumentList.dropFirst(startAt) {
+      guard var closure = arg.expression.as(ClosureExprSyntax.self) else {
+        closures.removeAll()
+        continue
+      }
+
+      // Trailing comma won't exist any more, move its trivia to the end of
+      // the closure instead
+      if let comma = arg.trailingComma {
+        closure = closure.with(\.trailingTrivia, closure.trailingTrivia.merging(triviaOf: comma))
+      }
+      closures.append((arg.label ?? .wildcardToken(), closure))
+    }
+
+    guard !closures.isEmpty else {
+      return nil
+    }
+
+    let trailingClosure = closures.first!.closure
+    let additionalTrailingClosures = closures.dropFirst().map {
+      MultipleTrailingClosureElementSyntax(
+        label: $0.label,
+        colon: .colonToken(),
+        closure: $0.closure
+      )
+    }
+
+    var converted = self.detach()
+
+    // Remove parens if there's no non-closure arguments left. Just remove the
+    // last comma otherwise and move its trivia to after the right paren (ie.
+    // before the first trailing closure).
+    var argList = Array(argumentList.dropLast(closures.count))
+    if argList.isEmpty {
+      converted =
+        converted
+        .with(\.leftParen, nil)
+        .with(\.rightParen, nil)
```

Comment:
Again, dropping the trivia of the `leftParen` and `rightParen`. Should we attach it to the first trailing closure instead? I wonder whether it would make sense to have a function like the following that removes a token and moves its trailing trivia to adjacent tokens. That way you wouldn’t really need to worry about moving the trivia for the tokens you’re removing somewhere. I haven’t tested it thoroughly though.

```swift
class TransferTrivia: SyntaxRewriter {
  var tokenToRemove: TokenSyntax

  init(tokenToRemove: TokenSyntax) {
    self.tokenToRemove = tokenToRemove
  }

  override func visitAny(_ node: Syntax) -> Syntax? {
    if !node.byteRange.intersects(tokenToRemove.byteRange) {
      // Nothing to rewrite. Skip this subtree
      return node
    }
    return nil
  }

  override func visit(_ token: TokenSyntax) -> TokenSyntax {
    if token.nextToken(viewMode: .sourceAccurate) == tokenToRemove {
      let leadingTriviaToTransfer = (tokenToRemove.leadingTrivia + tokenToRemove.trailingTrivia).prefix(while: { !$0.isNewline })
      return token.with(\.trailingTrivia, token.trailingTrivia + Trivia(pieces: leadingTriviaToTransfer))
    } else if token.previousToken(viewMode: .sourceAccurate) == tokenToRemove {
      let trailingTriviaToTransfer = (tokenToRemove.leadingTrivia + tokenToRemove.trailingTrivia).drop(while: { !$0.isNewline })
      return token.with(\.leadingTrivia, token.leadingTrivia + Trivia(pieces: trailingTriviaToTransfer))
    } else {
      return token
    }
  }
}

extension SyntaxProtocol {
  func removingRetainingTrivia(_ keyPath: WritableKeyPath<Self, TokenSyntax?>) -> Self {
    guard let token = self[keyPath: keyPath] else {
      return self
    }
    let transferredTrivia = TransferTrivia(tokenToRemove: token).rewrite(Syntax(self)).cast(Self.self)
    return transferredTrivia.with(keyPath, nil)
  }
}
```

Reply:
It's only dropping `leftParen`, `rightParen` is added below. I don't think adding to the next token is quite what we want, ie. I would expect the trivia of `)` to end up after the call.